### ISSUE TITLE
Packets from Pods in a specific node  end up with node IP as source

### ISSUE THREAD

**Issue by javiku** [projectcalico/calico#3440]
Hi, we are experiencing a very strange behaviour in one of our Kubernetes clusters.

We started detecting that Pods that where scheduled in a specific node (let's call it node X) failed to communicate with Pods in other nodes they had permissions to, according to the valid Network Nolicies in place. Whenever the Pods were in other nodes, communications worked as expected.

After some exhaustive digging (tcpdump, iptables), we have found out that **packets sent from the Pods in node X, are received in other nodes with node X host IP as source, instead of the Pod IP.**. Since Network Policies enforcement is based on the source pod IP, the packets are rejected.

We have managed to reproduce the same behaviour later with pods not subject to Network Policies.

#### Normal working scenario (sending pod not in node X)

- Sending Pod IP: `172.26.50.28`
- Destination Pod IP: `172.16.1.164`

Tcpdump at source **calico** interface (`calixxxxxxxx`)
```
IP 172.26.50.28.54990 > 172.16.1.164.80: Flags [S], seq 2293559199, win 64400, options [mss 1400,sackOK,TS val 3540383985 ecr 0,nop,wscale 7], length 0
```
Tcpdump at destination **calico** interface (`calixxxxxxxx`)
```
IP 172.26.50.28.54990 > 172.16.1.164.80: Flags [S], seq 2293559199, win 64400, options [mss 1400,sackOK,TS val 3540383985 ecr 0,nop,wscale 7], length 0
```

#### Failing scenario (sending pod is in node X)

- Sending Pod IP: `172.17.178.96`
- Destination Pod IP: `172.16.1.164`
- Node X host IP: `10.0.18.31`
- Node X `tunl0` IP: `172.17.178.110`

Tcpdump at source **calico** interface (`calixxxxxxxx`)
```
IP 172.17.178.110.56374 > 172.16.1.164.80: Flags [S], seq 2822404940, win 64400, options [mss 1400,sackOK,TS val 3287956126 ecr 0,nop,wscale 7], length 0
```
**Note that source IP is set to the `tunl0` interface IP (!)**.

Tcpdump at destination **calico** interface (`calixxxxxxxx`)
```
IP 10.0.18.31.56374 > 172.16.1.164.80: Flags [S], seq 2822404940, win 64400, options [mss 1400,sackOK,TS val 3287956126 ecr 0,nop,wscale 7], length 0
```

## Expected Behavior

Packets from other Pods received with the sending Pod IP as source.

## Current Behavior

For some unknown reason, packets sent from Pods running in a specific node, end up with the node's IP as source.
This breaks Network Policies enforcement, since they are based in packets source IPs.

## Your Environment

Calico version: 3.8.8
Orchestrator version : Kubernetes 1.16
Operating System and version: Ubuntu 18.04

**Comment by caseydavenport** [comment]
Sounds like the packets are probably getting NAT'd for some reason.

I would check the output from the following commands to see if any NAT rules are getting hit on the source node:

```
sudo iptables-save -c  | grep SNAT
```

```
sudo iptables-save -c | grep MASQUERADE 
```

If you see packet counts there it might be an indication of why traffic is being SNAT'd (if that's the root cause). 

Other things to check:
- Do you have multiple IP pools? 
- Do the IP pools you are using have NatOutgoing: true set?
- How did you install Calico?

**Comment by javiku** [comment]
Thank you @caseydavenport  for your tips, actually I was able to find the error source yesterday, a few hours before your reply..

The problem was that our configured 172.16.0.0/12 Pods CIDR was in conflict with Docker bridge CIDR, automatically set to 172.17.0.0/16. One of the standard Docker iptables rules was responsible for NATing packets from the node (or nodes, see comment at the end) where the assigned IP Block was in that range:
```
-A POSTROUTING -s 172.17.0.0/16 ! -o docker0 -j MASQUERADE
```
Deleting that rule for testing, immediately resulted in everything working as expected. Restarting the cluster with non-overlapping CIDRs obviously worked too.

Still, two comments:

1. I really miss this being **documented as a common pitfall**. After finding the problem I googled this specific topic and found very little references, some Azure AKS documentation, a comment from you "hidden" in [another issue](https://github.com/projectcalico/k8s-exec-plugin/issues/34#issuecomment-142430806) stating:

>  *"One thing I did not mention above is that when enabling Calico IPAM, you should make sure that your Docker bridge (cbr0) is not configured with an IP address within the Calico IP pool. This prevents conflicts between the docker bridge (which is still required for running docker) and the IPs assigned by Calico.*

2. In my testing I found the fact that a node was always assigned an IP Block very similar to the one assigned in previous clusters (or even the exact same one). In my scenario, worker X was always the only one assigned with a 172.17.x.x/26 conflicting IP block.
**How does this apparent determinism work?** Is there any documentation about how the IP Block assignment to nodes work? I read about selectors and such, but the default mechanism is not mentioned.

Thank you very much for the help. Cheers!
